lib/burnside/intrinsics.cc

```diff
@@ -14,85 +14,421 @@
 
 #include "intrinsics.h"
 #include "builder.h"
+#include "fir/FIROps.h"
+#include <unordered_map>
+#include <utility>
 
 /// [Coding style](https://llvm.org/docs/CodingStandards.html)
 
 namespace Fortran::burnside {
 
-// Define a simple static runtime description that will be transformed into
-// IntrinsicImplementation when building the IntrinsicLibrary.
-namespace runtime {
-enum class Type { f32, f64 };
-struct StaticDescription {
-  const char *name;
-  const char *symbol;
+/// MathRuntimeLibrary maps Fortran generic intrinsic names to runtime function
+/// signatures. There is no guarantee that that runtime functions are available
+/// for all intrinsic functions and possible types.
+/// To be easy and fast to use, this class holds a map and uses
+/// mlir::FunctionType to represent the runtime function type. This imply that
+/// MathRuntimeLibrary cannot be constexpr built and requires an
+/// mlir::MLIRContext to be built. Its constructor uses a constexpr table
+/// description of the runtime. The runtime functions are not declared into the
+/// mlir::module until there is a query that needs them. This is to avoid
+/// polluting the FIR/LLVM IR dumps with unused functions.
+class MathRuntimeLibrary {
+public:
+  /// The map key are Fortran generic intrinsic names.
+  using Key = llvm::StringRef;
+  struct Hash {  // Need custom hash for this kind of key
+    size_t operator()(const Key &k) const { return llvm::hash_value(k); }
+  };
+  /// Runtime function description that is sufficient to build an
+  /// mlir::FuncOp and to compare function types.
+  struct RuntimeFunction {
+    RuntimeFunction(llvm::StringRef n, mlir::FunctionType t)
+      : symbol{n}, type{t} {};
+    llvm::StringRef symbol;
+    mlir::FunctionType type;
+  };
+  using Map = std::unordered_multimap<Key, RuntimeFunction, Hash>;
```

Comment:
Thanks for pointing to this, after reading llvm do, I think I cannot use it here because I want a multimap and not a map. However, llvm doc points out that, very often, custom implementation based on sorted vectors are better than `std::multimap`. I do not want to start creating custom storage class just for performance sake yet, but I will keep this in mind.